posts/Gunnar/2019-01-23-testing-cdi-beans-and-persistence-layer-under-java-se.adoc

```diff
@@ -0,0 +1,456 @@
+plain unit tests of single classes to comprehensive integration tests deployed into a container (e.g. via http://arquillian.org/[Arquillian])
+and driven through tools such as http://rest-assured.io/[REST Assured].
+
+In this post I'd like to discuss one testing approach which represents some kind of a middle ground:
+launching a local CDI container and a JPA runtime, connected to an in-memory database.
+That way you can test CDI beans (e.g. containing business logic) in conjunction with the persistence layer (e.g. JPA-based repositories) under plain Java SE.
+
+This allows to test individual classes and components as they interact with others
+(e.g. without mocking of repositories when testing the business logic),
+while still benefitting from fast execution times (no container management/deployment and remote API calls needed).
+The approach also allows tests around services our application might rely on, such as interceptors, events, transactional semantics and other things
+that'd otherwise require deployment into a container.
+Finally, these tests are easy to debug, as everything runs in the local VM and no remote processes are involved.
+
+To make the approach worthwhile, the following requirements should be satisfied by the testing infrastructure:
+
+* Allow to obtain CDI beans via dependency injection, with support for all the CDI goodness such as interceptors, decorators, events etc.
+* Allow to obtain a JPA entity manager via dependency injection
+* Allow to use dependency injection in JPA entity listeners
+* Allow to use declarative transaction control via `@Transactional`
+* Allow to use transactional event observers (e.g. event observers running after transaction completion)
+
+In the following let's see how these requirements can be addressed.
+You can find a complete version of the shown code in the Hibernate https://github.com/hibernate/hibernate-demos/tree/master/other/cdi-jpa-testing[examples repository] on GitHub.
+That example project uses https://weld.cdi-spec.org/[Weld] as the CDI container, https://hibernate.org/orm/[Hibernate ORM] as the JPA provider and http://www.h2database.com/html/main.html[H2] as a database.
+Note the post is mostly interested in the interaction of CDI and the persistence layer, you could also use this approach with any other database such as Postgres or MySQL.
+
+== Obtaining CDI Beans via Dependency Injection
+
+Firing up a CDI container under Java SE is trivial using the http://www.cdi-spec.org/news/2017/05/15/CDI_2_is_released/[bootstrap API] standardized in CDI 2.0.
+So we could simply go and use that API in our tests.
+One alternative to consider is https://github.com/weld/weld-junit[Weld JUnit], a small extension to Weld (the CDI reference implementation) aimed at the purposes of testing.
+Amongst other things, Weld JUnit allows to specify which CDI scopes should be active when invoking CDI beans during the test.
```

Comment:
Might be worth mentioning that it also allows you to resolve injection into test instance (and does that automatically for JUnit 5) if only because you use that in the following code sample.

Reply:
Done.